### PR TITLE
Fix regression & enable ClearInitLocals in System.Text.RegularExpressions

### PR DESCRIPTION
**Pull request by ViktorHofer**
I expect we will use more stackalloc in Regex over time therefore it makes sense to enable the switch now.

Relates to https://github.com/dotnet/corefx/pull/26993

Also fixes a bad code refactoring introduced by https://github.com/dotnet/corefx/pull/26877

### CODE REVIEW

**Comment by ViktorHofer** [comment]
ClearInitLocals disabled:

|     Method |     Mean |     Error |    StdDev |     Gen 0 |  Allocated |
|----------- |---------:|----------:|----------:|----------:|-----------:|
|  RegexCtor | 3.134 ms | 0.0647 ms | 0.1845 ms | 2058.5938 |   897.5 KB |
| RegexMatch | 4.174 ms | 0.0831 ms | 0.2007 ms | 2367.1875 | 1008.85 KB |

ClearInitLocals enabled:

|     Method |     Mean |     Error |    StdDev |     Gen 0 | Allocated |
|----------- |---------:|----------:|----------:|----------:|----------:|
|  RegexCtor | 3.082 ms | 0.0658 ms | 0.1931 ms | 2062.5000 | 886.42 KB |
| RegexMatch | 4.233 ms | 0.0842 ms | 0.2291 ms | 2367.1875 | 983.19 KB |

The benchmarks are kind of unstable, I wouldn't interpret anything based on the mean number but the allocation is very intersting. I didn't expect that clearing init locals impacts allocation.

**Comment by ViktorHofer** [comment]
@dotnet-bot test Windows x64 Debug Build
@dotnet-bot test Windows x64 Release Build

**Comment by ViktorHofer** [comment]
@dotnet-bot test Windows x86 Release Build

**Comment by stephentoub** [comment]
> Fix regression

Are we missing a test such that we didn't catch it in CI?

File: src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/ValueListBuilder.cs [hunk 1]
```diff
@@ -0,0 +1,23 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (stephentoub) [inline src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/ValueListBuilder.cs]: Can we rename the file to something like `ValueListBuilder.Pop.cs`?  The general convention we have with these partial files is that they're suffixed with the specialization they provide.

**Comment by ViktorHofer** [comment]
> Are we missing a test such that we didn't catch it in CI?

Perf regression as I created the struct wrapper twice...

**Comment by stephentoub** [comment]
> Perf regression as I created the struct wrapper twice

Ah :smile: Ok, thanks.

**Comment by ViktorHofer** [comment]
@dotnet-bot test NETFX x86 Release Build

**Comment by ViktorHofer** [comment]
@dotnet-bot test OSX x64 Debug Build
@dotnet-bot test Windows x64 Debug Build

**Pull request merged**